v1/tarball/image.go

```diff
@@ -15,17 +15,203 @@
-// // image accesses an image from a remote registry
-// type image struct{}
-// var _ v1.Image = (*image)(nil)
+type tarBlob struct {
+	name string
+}
+
+type image struct {
+	path     string
+	manifest *v1.Manifest
+	config   *v1.ConfigFile
+}
+
+var _ v1.Image = (*image)(nil)
 
 // Image exposes an image from the tarball at the provided path.
 func Image(p string) (v1.Image, error) {
-	return nil, fmt.Errorf("NYI: tarball.Image(%v)", p)
+	img := image{
+		path: p,
+	}
+	return &img, nil
+}
+
+func (i *image) FSLayers() ([]v1.Hash, error) {
+	panic("not implemented")
+}
+
+func (i *image) DiffIDs() ([]v1.Hash, error) {
+	panic("not implemented")
+}
+
+func (i *image) ConfigName() (v1.Hash, error) {
+	panic("not implemented")
+}
+
+func (i *image) BlobSet() (map[v1.Hash]struct{}, error) {
+	panic("not implemented")
+}
+
+func (i *image) Digest() (v1.Hash, error) {
+	panic("not implemented")
+}
+
+func (i *image) MediaType() (types.MediaType, error) {
+	panic("not implemented")
+}
+
+// Manifest returns the v1.Manifest for the specified image.
+// This method memoizes the result, avoiding repeated reads from the tarball.
+func (i *image) Manifest() (*v1.Manifest, error) {
+	if i.manifest == nil {
+		if err := i.loadManifest(); err != nil {
+			return nil, err
+		}
+	}
+	return i.manifest, nil
+}
+
+// singleManifest is the struct used inside the `manifest.json` file of a `docker save` tarball.
+type singleManifest struct {
+	Config   string
+	RepoTags []string
+	Layers   []string
+}
+
+// tarManifest is the struct used to represent a single image inside a `docker save` tarball.
+type tarManifest []singleManifest
+
+func (i *image) loadManifest() error {
+	mfstBytes, err := extractFileFromTar(i.path, "manifest.json")
+	if err != nil {
+		return err
+	}
+
+	tarMfst := tarManifest{}
+	if err := json.Unmarshal(mfstBytes, &tarMfst); err != nil {
+		return err
+	}
+	if len(tarMfst) != 1 {
+		return errors.New("tarball must contain only a single image to be used with tarball.Image")
+	}
+
+	cfgBytes, err := extractFileFromTar(i.path, tarMfst[0].Config)
+	if err != nil {
+		return err
+	}
+
+	cfg, err := v1.ParseConfigFile(cfgBytes)
+	if err != nil {
+		return err
+	}
+	i.config = cfg
+
+	manifest := v1.Manifest{
+		SchemaVersion: 2,
+		MediaType:     types.DockerManifestSchema2,
+		Config: v1.Descriptor{
+			MediaType: types.DockerConfigJSON,
+			Size:      int64(len(cfgBytes)),
+			Digest:    v1.SHA256(string(cfgBytes)),
+		},
+	}
+
+	for _, l := range tarMfst[0].Layers {
+		// TODO(dlorenc): support compressed layers.
+		uncompressed, err := extractFileFromTar(i.path, l)
+		if err != nil {
+			return err
+		}
+		layer, err := compress(uncompressed)
+		if err != nil {
+			return err
+		}
+
+		manifest.Layers = append(manifest.Layers, v1.Descriptor{
+			MediaType: types.DockerLayer,
+			Size:      int64(len(layer)),
+			Digest:    v1.SHA256(string(layer)),
+		})
+	}
+	i.manifest = &manifest
+	return nil
+}
+
+func compress(l []byte) ([]byte, error) {
```

Comment:
I think we want to switch to io.Reader everywhere to reduce our in-memory requirements.  This is a significant limitation of the python client (born of non-streaming http interfaces).